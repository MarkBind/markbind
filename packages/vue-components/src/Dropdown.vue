<template>
  <li
    v-if="isLi"
    ref="dropdown"
    :class="[{ 'disabled': disabledBool },
             isLi ? 'dropdown' : 'btn-group', addClass]"
  >
    <slot name="button">
      <a
        class="dropdown-toggle"
        role="button"
        :class="{'disabled': disabledBool}"
        @keyup.esc="hideDropdownMenu()"
      >
        <slot name="_header">
          <slot name="header"></slot>
        </slot>
      </a>
    </slot>
    <slot name="dropdown-menu" :class="[{ 'show': show }, { 'dropdown-menu-right': menuAlignRight }]">
      <ul class="dropdown-menu" :class="[{ 'show': show }, { 'dropdown-menu-right': menuAlignRight }]">
        <slot></slot>
      </ul>
    </slot>
  </li>
  <submenu v-else-if="isSubmenu" ref="submenu">
    <template v-for="(node, name) in $slots" :slot="name">
      <slot :name="name"></slot>
    </template>
  </submenu>
  <div
    v-else
    ref="dropdown"
    :class="[{ 'disabled': disabledBool },
             isLi ? 'dropdown' : 'btn-group', addClass]"
  >
    <slot name="before"></slot>
    <slot name="button">
      <button
        type="button"
        class="btn dropdown-toggle"
        :class="[btnType, btnWithBefore]"
        :disabled="disabledBool"
        @keyup.esc="hideDropdownMenu()"
      >
        <slot name="_header">
          <slot name="header"></slot>
        </slot>
      </button>
    </slot>
    <slot name="dropdown-menu" :class="[{ 'show': show }, { 'dropdown-menu-right': menuAlignRight }]">
      <ul class="dropdown-menu" :class="[{ 'show': show }, { 'dropdown-menu-right': menuAlignRight }]">
        <slot></slot>
      </ul>
    </slot>
  </div>
</template>

<script>
import { toBoolean } from './utils/utils';
import $ from './utils/NodeList';

export default {
  props: {
    disabled: {
      type: Boolean,
      default: false,
    },
    type: {
      type: String,
      default: 'light',
    },
    menuAlignRight: {
      type: Boolean,
      default: false,
    },
    addClass: {
      type: String,
      default: '',
    },
  },
  provide: { hasParentDropdown: true },
  inject: ['hasParentDropdown'],
  data() {
    return {
      show: false,
    };
  },
  computed: {
    btnType() {
      return `btn-${this.type}`;
    },
    disabledBool() {
      return toBoolean(this.disabled);
    },
    isLi() { return this.$parent._navbar || this.$parent.menu || this.$parent._tabset; },
    isSubmenu() { return this.hasParentDropdown; },
    menu() {
      return !this.$parent || this.$parent.navbar;
    },
    submenu() {
      return this.$parent && (this.$parent.menu || this.$parent.submenu);
    },
    slots() {
      return this.$slots.default;
    },
    btnWithBefore() {
      if (this.$slots.before) {
        return 'btn-with-before';
      }
      return '';
    },
  },
  methods: {
    blur() {
      this.unblur();
      this._hide = setTimeout(() => {
        this._hide = null;
        this.hideDropdownMenu();
      }, 100);
    },
    unblur() {
      if (this._hide) {
        clearTimeout(this._hide);
        this._hide = null;
      }
    },
    hideDropdownMenu() {
      this.show = false;
      $(this.$refs.dropdown).findChildren('ul').each(ul => ul.classList.toggle('show', false));
    },
    showDropdownMenu() {
      this.show = true;
      $(this.$refs.dropdown).findChildren('ul').each(ul => ul.classList.toggle('show', true));
    },
  },
  mounted() {
    const $el = $(this.$refs.dropdown);
    if (this.show) {
      this.showDropdownMenu();
    }
    $el.onBlur(() => { this.hideDropdownMenu(); }, false);
    $el.findChildren('a,button.dropdown-toggle').on('click', (e) => {
      e.preventDefault();
      if (this.disabledBool) { return false; }
      if (this.show) {
        this.hideDropdownMenu();
      } else {
        this.showDropdownMenu();
      }
      return false;
    });
    $el.findChildren('ul').on('click', 'li>a', (e) => {
      if (e.target.classList.contains('submenu-toggle')) { return; }
      this.hideDropdownMenu();
    });
  },
  beforeDestroy() {
    const $el = $(this.$refs.dropdown);
    $el.offBlur();
    $el.findChildren('a,button').off();
    $el.findChildren('ul').off();
  },
};
</script>

<style scoped>
    .secret {
        position: absolute;
        clip: rect(0 0 0 0);
        overflow: hidden;
        margin: -1px;
        height: 1px;
        width: 1px;
        padding: 0;
        border: 0;
    }

    .btn-with-before {
        padding-left: 0.2rem;
        padding-right: 0.4rem;
    }

    .dropdown-toggle {
        cursor: pointer;
        color: inherit;
        text-decoration: none;
    }
</style>
