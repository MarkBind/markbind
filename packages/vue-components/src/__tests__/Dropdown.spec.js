import { mount } from '@vue/test-utils';
import Dropdown from '../Dropdown.vue';

const DEFAULT_STUBS = {
  'dropdown': Dropdown,
};

const DROPDOWN = `
<li><a href="#" class="dropdown-item">Dropdown Item 1</a></li>
<li><a href="#" class="dropdown-item">Dropdown Item 2</a></li>
<li><a href="#" class="dropdown-item">Dropdown Item 3</a></li>
`;

const FIRST_LEVEL_NESTED_DROPDOWN = `
${DROPDOWN}
<dropdown header="Submenu Level 1">
  <li><a href="#" class="dropdown-item">Nested Dropdown Item 1</a></li>
  <li><a href="#" class="dropdown-item">Nested Dropdown Item 2</a></li>
</dropdown>
`;

const SECOND_LEVEL_NESTED_DROPDOWN = `
<dropdown header="Submenu Level 1">
  <li><a href="#" class="dropdown-item">Nested Dropdown Item 1</a></li>
  <li><a href="#" class="dropdown-item">Nested Dropdown Item 2</a></li>
  <dropdown header="Submenu Level 2">
    <li><a href="#" class="dropdown-item">Nested Submenu Item</a></li>
  </dropdown>
</dropdown>
`;

describe('Dropdown with submenu', () => {
  test('dropdown', async () => {
    const wrapper = mount(Dropdown, {
      slots: {
        header: 'Test Dropdown',
        default: DROPDOWN,
      },
      global: {
        stubs: DEFAULT_STUBS,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test('dropdown with one level of submenu', async () => {
    const wrapper = mount(Dropdown, {
      slots: {
        header: 'Test Dropdown',
        default: FIRST_LEVEL_NESTED_DROPDOWN,
      },
      global: {
        stubs: DEFAULT_STUBS,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test('dropdown with two levels of submenu', async () => {
    const wrapper = mount(Dropdown, {
      slots: {
        header: 'Test Dropdown',
        default: SECOND_LEVEL_NESTED_DROPDOWN,
      },
      global: {
        stubs: DEFAULT_STUBS,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
