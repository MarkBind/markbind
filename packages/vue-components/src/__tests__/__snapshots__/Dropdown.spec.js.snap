// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown with submenu dropdown 1`] = `
<div
  class="btn-group"
>
   
  <button
    class="btn dropdown-toggle btn-light"
    type="button"
  >
    Test Dropdown
  </button>
   
  <ul
    class="dropdown-menu"
  >
    <li>
      <a
        class="dropdown-item"
        href="#"
      >
        Dropdown Item 1
      </a>
    </li>
     
    <li>
      <a
        class="dropdown-item"
        href="#"
      >
        Dropdown Item 2
      </a>
    </li>
     
    <li>
      <a
        class="dropdown-item"
        href="#"
      >
        Dropdown Item 3
      </a>
    </li>
  </ul>
</div>
`;

exports[`Dropdown with submenu dropdown with one level of submenu 1`] = `
<div
  class="btn-group"
>
   
  <button
    class="btn dropdown-toggle btn-light"
    type="button"
  >
    Test Dropdown
  </button>
   
  <ul
    class="dropdown-menu"
  >
    <li>
      <a
        class="dropdown-item"
        href="#"
      >
        Dropdown Item 1
      </a>
    </li>
     
    <li>
      <a
        class="dropdown-item"
        href="#"
      >
        Dropdown Item 2
      </a>
    </li>
     
    <li>
      <a
        class="dropdown-item"
        href="#"
      >
        Dropdown Item 3
      </a>
    </li>
     
    <li
      class="dropdown-submenu dropright"
      haeder="Submenu Level 1"
    >
      <a
        class="submenu-toggle"
        role="button"
      />
       
      <ul
        class="dropdown-menu"
      >
        <li>
          <a
            class="dropdown-item"
            href="#"
          >
            Nested Dropdown Item 1
          </a>
        </li>
         
        <li>
          <a
            class="dropdown-item"
            href="#"
          >
            Nested Dropdown Item 2
          </a>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`Dropdown with submenu dropdown with two levels of submenu 1`] = `
<div
  class="btn-group"
>
   
  <button
    class="btn dropdown-toggle btn-light"
    type="button"
  >
    Test Dropdown
  </button>
   
  <ul
    class="dropdown-menu"
  >
    <li
      class="dropdown-submenu dropright"
      haeder="Submenu Level 1"
    >
      <a
        class="submenu-toggle"
        role="button"
      />
       
      <ul
        class="dropdown-menu"
      >
        <li>
          <a
            class="dropdown-item"
            href="#"
          >
            Nested Dropdown Item 1
          </a>
        </li>
         
        <li>
          <a
            class="dropdown-item"
            href="#"
          >
            Nested Dropdown Item 2
          </a>
        </li>
         
        <li
          class="dropdown-submenu dropright"
          haeder="Submenu Level 2"
        >
          <a
            class="submenu-toggle"
            role="button"
          />
           
          <ul
            class="dropdown-menu"
          >
            <li>
              <a
                class="dropdown-item"
                href="#"
              >
                Nested Submenu Item
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
